bob/src/hw_metrics_collector.rs

```diff
@@ -36,29 +36,68 @@ impl HWMetricsCollector {
+    /// correct mount point with disk path. If it fails to match, the function fallbacks to the first found mount point.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the function couldn't receive metadata from OS
     fn collect_used_disks(disks: &[DiskPath]) -> HashMap<PathBuf, DiskName> {
-        System::new_all()
-            .disks()
+        let sys_info = System::new_with_specifics(RefreshKind::new().with_disks_list());
+        let sys_metadata = Self::collect_metadata(sys_info.disks(), disks);
+
+        disks
             .iter()
-            .filter_map(|d| {
-                let path = d.mount_point();
-                disks
-                    .iter()
-                    .find(move |dp| {
-                        let dp_md = Path::new(dp.path())
-                            .metadata()
-                            .expect("Can't get metadata from OS");
-                        let p_md = path.metadata().expect("Can't get metadata from OS");
-                        p_md.dev() == dp_md.dev()
-                    })
-                    .map(|config_disk| {
-                        let diskpath = path.to_str().expect("Not UTF-8").to_owned();
-                        (PathBuf::from(diskpath), config_disk.name().clone())
-                    })
+            .filter_map(|disk| {
+                let disk_metadata = Path::new(disk.path())
+                    .metadata()
+                    .expect("Can't get metadata from OS");
+                sys_metadata
+                    .get(&disk_metadata.dev())
+                    .map(|sys_disk| (PathBuf::from(sys_disk.mount_point()), disk.name().clone()))
             })
             .collect()
     }
 
+    /// Maps disk's Inforamtion to the disk's device ID.
+    ///
+    /// The fucntion will try to match mount point with disk path
+    /// If it's fails to do so, it will fallback to the first found mount point for the specified
+    /// device ID
+    ///
+    /// # Panics
+    ///
+    /// Panics if the function couldn't receive metadata from OS
+    fn collect_metadata<'a>(sys_disks: &'a [Disk], disks: &[DiskPath]) -> HashMap<u64, &'a Disk> {
+        let dev_path: Vec<(_, _)> = disks
+            .iter()
+            .map(|disk| {
+                let path = Path::new(disk.path());
+                let disk_metadata = path.metadata().expect("Can't get metadata from OS");
+                (disk_metadata.dev(), path)
+            })
+            .collect();
+        let mut res = HashMap::new();
+        for disk in sys_disks {
+            let metadata = disk
+                .mount_point()
+                .metadata()
+                .expect("Can't get metadata from OS");
+            res.entry(metadata.dev()).or_insert(disk);
+            if dev_path
+                .iter()
+                .filter(|(dev, _)| *dev == metadata.dev())
+                .any(|(_, path)| disk.mount_point().file_name() == path.file_name())
```

Comment:
This is not correct. We should check that `disk.mount_point()` is a prefix of `path`. Example: `mount_point = "/opt/disk1/", path = "/opt/disk1/bob"`. Use `starts_with` function